src/Microsoft.AspNetCore.Routing.Abstractions/MetadataCollection.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class EndpointMetadataCollection : IReadOnlyList<object>
+    {
+        public static readonly EndpointMetadataCollection Empty = new EndpointMetadataCollection(Array.Empty<object>());
+
+        private readonly object[] _items;
+
+        public EndpointMetadataCollection(IEnumerable<object> items)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            _items = items.ToArray();
+        }
+
+        public object this[int index] => _items[index];
+
+        public int Count => _items.Length;
+
+        public T GetMetadata<T>() where T : class
+        {
+            for (var i = _items.Length -1; i >= 0; i--)
+            {
+                var item = _items[i] as T;
+                if (item !=null)
+                {
+                    return item;
+                }
+            }
+
+            return default;
+        }
+
+        public IEnumerable<T> GetOrderedMetadata<T>() where T : class
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i] as T;
+                if (item != null)
+                {
+                    yield return item;
+                }
+            }
+        }
```

Comment:
@kichalla - this will look somewhat familiar to you. It's a parallel of https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs#L84

Reply:
We can polish this more in the future as we actually us it.

Reply:
Yup 👍